api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -499,10 +499,14 @@ Builder add(int sourceId, int fieldId, String name, Transform<?, ?> transform) {
     }
 
     public PartitionSpec build() {
-      PartitionSpec spec = new PartitionSpec(schema, specId, fields, lastAssignedFieldId.get());
+      PartitionSpec spec = buildUnchecked();
       checkCompatibility(spec, schema);
       return spec;
     }
+
+    PartitionSpec buildUnchecked() {
+      return new PartitionSpec(schema, specId, fields, lastAssignedFieldId.get());
+    }
```

Comment:
This is needed for table replacement. Previously, building the replacement table metadata carried through partition specs and sort orders unchanged. Now that replace table goes through the same method to update the schema, `setCurrentSchema`, the specs and orders are now rebuilt to reference the current schema.

When replacing a table, its schema, partition spec, and sort order only need to be consistent with one another, not the older specs and sort orders. This relaxes the check so that tests that have incompatible changes don't fail.

Relaxing checks is the correct thing to do because consistency is now checked in the `build` method. It allows the current schema, spec, and order to be changed as long as they are consistent with one another. Older specs and orders are still kept because they may be referenced by data files in existing manifests.